### DISCUSSION TITLE
Any AMD user successfully use this fork with directML

### DISCUSSION THREAD

**Discussion by howDoILearnPython** [vladmandic/sdnext#1021] [General]
i noticed the development update on 13/5 said the directML support is enhanced.  i tried git pull to the latest version and start webui with --use-directML.  There are INFO said it's using DirectML Backend.  However, the computer still use CPU in image generation.  

Is there any AMD user here?  any successful experience can share?

**Comment by vladmandic** [comment]
there are plenty of amd users here. what did it actually say on console - don't just say "there is no info"

**Comment by howDoILearnPython** [comment]
just try repeat with --upgrade flag...
this time i got exception as below.  seems i need to delete everything and reinstall.  (it take time....)

Using VENV: C:\vladmandic\venv
19:59:57-114311 INFO     Starting SD.Next
19:59:57-114311 INFO     Python 3.10.9 on Windows
19:59:57-239306 INFO     Version: 6c271dcf Wed May 17 07:16:11 2023 -0400
19:59:57-598855 INFO     Latest published version: 616553220b5992ac5f0e2222afd176c578bfc079 2023-05-17T11:42:41Z
19:59:57-598855 INFO     Setting environment tuning
19:59:57-598855 INFO     Using DirectML Backend
20:00:00-130732 INFO     Torch 2.0.0+cpu
20:00:00-130732 WARNING  Torch reports CUDA not available
20:00:00-130732 INFO     Verifying requirements
20:00:00-162359 INFO     Installing packages
20:00:00-224648 INFO     Running setup
20:00:00-224648 INFO     Installing repositories
20:00:00-458988 INFO     Installing submodules
20:00:03-115953 INFO     Updating submodules
20:00:41-610988 INFO     Extensions enabled: ['a1111-sd-webui-lycoris', 'clip-interrogator-ext', 'LDSR', 'Lora',
                         'multidiffusion-upscaler-for-automatic1111', 'prompt-bracket-checker', 'ScuNET',
                         'sd-dynamic-thresholding', 'sd-extension-aesthetic-scorer', 'sd-extension-steps-animation',
                         'sd-extension-system-info', 'sd-webui-controlnet', 'sd-webui-model-converter', 'seed_travel',
                         'stable-diffusion-webui-images-browser', 'stable-diffusion-webui-rembg', 'SwinIR']
20:00:41-626617 INFO     Updating Wiki
20:00:42-705102 INFO     Running extension preloading
20:00:42-720733 INFO     Server arguments: ['--use-directml']
No module 'xformers'. Proceeding without it.
Error in sys.excepthook:
Traceback (most recent call last):
  File "C:\vladmandic\venv\lib\site-packages\rich\traceback.py", line 103, in excepthook
    Traceback.from_exception(
  File "C:\vladmandic\venv\lib\site-packages\rich\traceback.py", line 346, in from_exception
    return cls(
  File "C:\vladmandic\venv\lib\site-packages\rich\traceback.py", line 280, in __init__
    for suppress_entity in suppress:
TypeError: 'module' object is not iterable

Original exception was:
Traceback (most recent call last):
  File "C:\vladmandic\launch.py", line 141, in <module>
    instance = start_server(immediate=True, server=None)
  File "C:\vladmandic\launch.py", line 116, in start_server
    module_spec.loader.exec_module(server)
  File "<frozen importlib._bootstrap_external>", line 883, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "C:\vladmandic\webui.py", line 29, in <module>
    from modules import import_hook # pylint: disable=W0611,C0411,C0412
  File "C:\vladmandic\modules\import_hook.py", line 2, in <module>
    from modules.shared import opts, log
  File "C:\vladmandic\modules\shared.py", line 613, in <module>
    devices.device, devices.device_interrogate, devices.device_gfpgan, devices.device_esrgan, devices.device_codeformer = (devices.cpu if any(y in cmd_opts.use_cpu for y in [x, 'all']) else devices.get_optimal_device() for x in ['sd', 'interrogate', 'gfpgan', 'esrgan', 'codeformer'])
  File "C:\vladmandic\modules\shared.py", line 613, in <genexpr>
    devices.device, devices.device_interrogate, devices.device_gfpgan, devices.device_esrgan, devices.device_codeformer = (devices.cpu if any(y in cmd_opts.use_cpu for y in [x, 'all']) else devices.get_optimal_device() for x in ['sd', 'interrogate', 'gfpgan', 'esrgan', 'codeformer'])
  File "C:\vladmandic\modules\devices.py", line 54, in get_optimal_device
    return torch.device(get_optimal_device_name())
  File "C:\vladmandic\modules\devices.py", line 42, in get_optimal_device_name
    import torch_directml # pylint: disable=import-error
ModuleNotFoundError: No module named 'torch_directml'

**Comment by howDoILearnPython** [comment]
wow.
May be that is my old installation problem?

My previous build was installed by simply launch webui.bat
and subsequently started with webui --use-directml

i donn't know if this matter, This time, i install it with webui --use-directml
and the console output is different.  Tried and it is using GPU now. 

20:26:27-093511 INFO     Starting SD.Next
20:26:27-093511 INFO     Python 3.10.9 on Windows
20:26:27-202909 INFO     Version: 61655322 Wed May 17 07:42:41 2023 -0400
20:26:28-703101 INFO     Setting environment tuning
20:26:28-703101 INFO     Using DirectML Backend
20:26:31-610135 INFO     Torch 2.0.0+cpu
20:26:31-625515 INFO     Torch backend: DirectML (torch-directml 0.2.0.dev230426)
**20:26:31-781679 INFO     Torch detected GPU: AMD Radeon(TM) Vega 8 Graphics
20:26:31-859820 INFO     Torch detected GPU: AMD Radeon(TM) Vega 8 Graphics**
20:26:31-859820 INFO     Installing package: tensorflow==2.12.0
20:27:29-289866 INFO     Verifying requirements
20:27:29-321124 WARNING  Package wrong version: timm 0.4.12 required 0.6.13
20:27:29-321124 INFO     Installing package: timm==0.6.13
20:27:35-525166 INFO     Installing packages
20:27:35-603307 INFO     No changes detected: Quick launch active
20:27:35-618936 INFO     Running extension preloading
20:27:35-618936 INFO     Server arguments: ['--use-directml']
No module 'xformers'. Proceeding without it.
20:27:47-308006 INFO     Libraries loaded
20:27:47-323634 INFO     Available models: C:\vladmandic\models\Stable-diffusion 1
ControlNet v1.1.173
ControlNet v1.1.173
Image Browser: ImageReward is not installed, cannot be used.

**Comment by vladmandic** [comment]
> My previous build was installed by simply launch webui.bat
> and subsequently started with webui --use-directml

Yes, once `torch` is installed, it will be used as-is. If you want to force reinstall of correct torch when you want to start using `--use-directml`, you can add `--reinstall` flag.

But since you've now done it with fresh install, its a a moot point.

**Reply by Aptronymist** [reply]
Yeah, I was gonna say I tried my dual-card setup today and after specifying both --use-directml and --device-id 0, it worked like a charm, but I did have to manually install torch-directml after receiving that same error. Thankfully I've been doing this stuff for months and knew what to do. I've yet to try running two copies of SD.Next at the same time (say for doing some light TI training AND diffusion at the same time), but I'll try that next and let you know.

Oh, as an afterthought, it might be useful for it to state which device id you're using, and others available, somewhere in the startup log.

**Reply by vladmandic** [reply]
It's already logged, just start with --debug flag.

**Comment by eightbitfit** [comment]
I'm having a similar issue as above, and have tried things menitoned here, including numerous clean installs. I have installed DirectML and confirmed I can use it per this documentation: https://learn.microsoft.com/en-us/windows/ai/directml/gpu-pytorch-windows.  Almos seems like a PATH issue, but I added anaconda into my path and used it to test torch.

PS D:\AI\Images\SD\stable-diffusion-webui-directml> .\webui --use-directml --debug
venv "D:\AI\Images\SD\stable-diffusion-webui-directml\venv\Scripts\Python.exe"
fatal: No names found, cannot describe anything.
Python 3.10.9 | packaged by Anaconda, Inc. | (main, Mar  1 2023, 18:18:15) [MSC v.1916 64 bit (AMD64)]
Version: <none>
Commit hash: 5e48f300c1b74f0312f692dc8d0485b5120fc497
Installing requirements
Launching Web UI with arguments: --use-directml --debug
Traceback (most recent call last):
  File "D:\AI\Images\SD\stable-diffusion-webui-directml\launch.py", line 38, in <module>
    main()
  File "D:\AI\Images\SD\stable-diffusion-webui-directml\launch.py", line 34, in main
    start()
  File "D:\AI\Images\SD\stable-diffusion-webui-directml\modules\launch_utils.py", line 325, in start
    import webui
  File "D:\AI\Images\SD\stable-diffusion-webui-directml\webui.py", line 42, in <module>
    from modules.call_queue import wrap_gradio_gpu_call, wrap_queued_call, queue_lock  # noqa: F401
  File "D:\AI\Images\SD\stable-diffusion-webui-directml\modules\call_queue.py", line 7, in <module>
    from modules import shared, progress
  File "D:\AI\Images\SD\stable-diffusion-webui-directml\modules\shared.py", line 11, in <module>
    import modules.interrogate
  File "D:\AI\Images\SD\stable-diffusion-webui-directml\modules\interrogate.py", line 14, in <module>
    from modules import devices, paths, shared, lowvram, modelloader, errors
  File "D:\AI\Images\SD\stable-diffusion-webui-directml\modules\devices.py", line 4, in <module>
    import torch_directml
ModuleNotFoundError: No module named 'torch_directml'
Press any key to continue . . .

**Reply by brknsoul** [reply]
After you git cloned the repository, the very first time you ran `webui.bat`, did you run it with `--use-directml`?

**Reply by eightbitfit** [reply]
I did not the first time, but after reading this thread I wiped it clean and did so. The post above is after that configuration.

**Reply by vladmandic** [reply]
start `webui --debug` and post first ~15 lines of the console log

**Reply by eightbitfit** [reply]
This is everything I get:

PS D:\AI\Images\SD> .\webui --debug
venv "D:\AI\Images\SD\venv\Scripts\Python.exe"
fatal: No names found, cannot describe anything.
Python 3.10.9 | packaged by Anaconda, Inc. | (main, Mar  1 2023, 18:18:15) [MSC v.1916 64 bit (AMD64)]
Version: <none>
Commit hash: 5e48f300c1b74f0312f692dc8d0485b5120fc497
Installing requirements
Launching Web UI with arguments: --debug
Traceback (most recent call last):
  File "D:\AI\Images\SD\launch.py", line 38, in <module>
    main()
  File "D:\AI\Images\SD\launch.py", line 34, in main
    start()
  File "D:\AI\Images\SD\modules\launch_utils.py", line 325, in start
    import webui
  File "D:\AI\Images\SD\webui.py", line 42, in <module>
    from modules.call_queue import wrap_gradio_gpu_call, wrap_queued_call, queue_lock  # noqa: F401
  File "D:\AI\Images\SD\modules\call_queue.py", line 7, in <module>
    from modules import shared, progress
  File "D:\AI\Images\SD\modules\shared.py", line 11, in <module>
    import modules.interrogate
  File "D:\AI\Images\SD\modules\interrogate.py", line 14, in <module>
    from modules import devices, paths, shared, lowvram, modelloader, errors
  File "D:\AI\Images\SD\modules\devices.py", line 4, in <module>
    import torch_directml
ModuleNotFoundError: No module named 'torch_directml'
Press any key to continue . . .
PS D:\AI\Images\SD>

**Reply by vladmandic** [reply]
this looks like installer barely started and then python decided to basically skip all that and move on to launching.
there are quite a few lines that should be printed between 

> Installing requirements
> Launching Web UI with arguments: --debug

and i just saw you're using 
> Python 3.10.9 | packaged by Anaconda, Inc. | (main, Mar 1 2023, 18:18:15) [MSC v.1916 64 bit (AMD64)]

why anaconda? this repo does not use or run in conda environment, it uses standard python and creates venv.
can you install regular python?

and this is not related at all to the issue that original author noted. once you follow recommendations and there are issues, please create a separate one so we don't hijaack focus from original author here.

**Reply by eightbitfit** [reply]
Some of the installs I have read recommended Anaconda for Windows, if that's not optimal I'll certainly change. I was looking at the torch_directml error as my key focus. Appreciate the guidance.

**Reply by eightbitfit** [reply]
I just removed anaconda, cleaned venv  and re-ran webui --debug. Still getting stuck on missing torch_directml


[notice] A new release of pip available: 22.2.1 -> 23.1.2
[notice] To update, run: D:\AI\Images\SD\venv\Scripts\python.exe -m pip install --upgrade pip
Installing gfpgan
Installing clip
Installing open_clip
Installing requirements for CodeFormer
Installing requirements
Launching Web UI with arguments: --debug
Traceback (most recent call last):
  File "D:\AI\Images\SD\launch.py", line 38, in <module>
    main()
  File "D:\AI\Images\SD\launch.py", line 34, in main
    start()
  File "D:\AI\Images\SD\modules\launch_utils.py", line 325, in start
    import webui
  File "D:\AI\Images\SD\webui.py", line 42, in <module>
    from modules.call_queue import wrap_gradio_gpu_call, wrap_queued_call, queue_lock  # noqa: F401
  File "D:\AI\Images\SD\modules\call_queue.py", line 7, in <module>
    from modules import shared, progress
  File "D:\AI\Images\SD\modules\shared.py", line 11, in <module>
    import modules.interrogate
  File "D:\AI\Images\SD\modules\interrogate.py", line 14, in <module>
    from modules import devices, paths, shared, lowvram, modelloader, errors
  File "D:\AI\Images\SD\modules\devices.py", line 4, in <module>
    import torch_directml
ModuleNotFoundError: No module named 'torch_directml'
Press any key to continue . . .

**Reply by vladmandic** [reply]
Open a separarate thread and post full log.
This is somehow not a valid python env, installer is not executing correctly. The fact that torch-directml is not found is just tip of the iceberg because it happens to be the first such package.
